java/vector/src/main/java/org/apache/arrow/vector/compare/ApproxEqualsVisitor.java

```diff
@@ -53,13 +51,63 @@ public ApproxEqualsVisitor(ValueVector left, ValueVector right, float floatEpsil
     super(left, right, true);
     this.floatEpsilon = floatEpsilon;
     this.doubleEpsilon = doubleEpsilon;
+
+    floatDiffFunction = (vector1, index1, vector2, index2) -> {
+      boolean isNull1 = vector1.isNull(index1);
+      boolean isNull2 = vector2.isNull(index2);
+
+      if (isNull1 || isNull2) {
+        if (isNull1 && isNull2) {
+          return true;
+        } else {
+          // exactly one of them is null
+          return false;
+        }
+      }
+
+      float leftValue = vector1.get(index1);
+      float rightValue = vector2.get(index2);
+
+      if (Float.isNaN(leftValue)) {
+        return Float.isNaN(rightValue);
+      }
+      if (Float.isInfinite(leftValue)) {
+        return Float.isInfinite(rightValue) && Math.signum(leftValue) == Math.signum(rightValue);
+      }
+      return Math.abs(leftValue - rightValue) <= floatEpsilon;
+    };
+
+    doubleDiffFunction = (vector1, index1, vector2, index2) -> {
```

Comment:
I think this change is too much, I think the right thing to do is introduce an interface is to change DiffFunction into two interface.  

DiffFunctionFloat {
   bool approxEquals(float v1, float v2);
}

DiffFunctionDouble {
  bool approxEquals(double v1, double v2);
}

and

Reply:
I see. Will revise accordingly.